### PR TITLE
schedule-service: add feign client for school-class-service and some …

### PR DESCRIPTION
**Pull request by vzhuleho**

### CODE REVIEW

File: scheduleservice/Dockerfile [hunk 1]
```diff
@@ -1,3 +1,3 @@
 FROM openjdk:11-jre-slim
-COPY build/libs/scheduleservice-0.0.1-SNAPSHOT.jar /opt
-ENTRYPOINT ["java","-jar","/opt/scheduleservice-0.0.1-SNAPSHOT.jar"]
\ No newline at end of file
+COPY scheduleservice.jar /opt
```
Comment (xpinjection) [inline scheduleservice/Dockerfile]: probably it is better to use build directory as a source for all built artefacts

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java [hunk 2]
```diff
@@ -65,4 +74,39 @@ public SubjectDetails getSubject(long id) {
 		List<SubjectDetails> getSubjects();
 
 	}
+
+	@RequiredArgsConstructor
+	private class ExternalSchoolClassDataProvider implements SchoolClassDataProvider {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java]: It is better to move such code in dedicated classes.

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java [hunk 3]
```diff
@@ -35,12 +39,17 @@ public CurriculumDataProvider curriculumDataProvider() {
 		return new ExternalCurriculumDataProvider();
 	}
 
+	@Bean
+	public SchoolClassDataProvider schoolClassDataProvider() {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java]: You could just mark class with Component annotation.

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java [hunk 4]
```diff
@@ -24,6 +27,7 @@
 
 	private final UserServiceClient userServiceClient;
 	private final CurriculumServiceClient curriculumServiceClient;
+	private final SchoolServiceClient schoolServiceClient;
 
 	@Bean
 	public UserDataProvider userDataProvider() {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java]: You could just mark class with Component annotation.

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java [hunk 5]
```diff
@@ -37,6 +43,11 @@ public CurriculumDataProvider curriculumDataProvider() {
 		return new LocalCurriculumDataProvider();
 	}
 
+	@Bean
+	public SchoolClassDataProvider schoolClassDataProvider() {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java]: You could just mark class with Component annotation.

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java [hunk 6]
```diff
@@ -62,4 +73,26 @@ public SubjectDetails getSubject(long id) {
 		}
 	}
 
+	private class LocalSchoolClassDataProvider implements SchoolClassDataProvider {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java]: Again it is better to use dedicated classes.

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/SchoolClassDataProvider.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package com.kyriba.school.scheduleservice.dataprovider;
+
+import com.kyriba.school.scheduleservice.domain.dto.PupilDetails;
+import com.kyriba.school.scheduleservice.domain.dto.SchoolClassDetails;
+
+import java.util.List;
+
+public interface SchoolClassDataProvider {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/SchoolClassDataProvider.java]: This is repository, no?

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/service/ScheduleNotFoundException.java [hunk 8]
```diff
@@ -2,7 +2,10 @@
 
 import org.springframework.data.rest.webmvc.ResourceNotFoundException;
 
-class ScheduleNotFoundException extends ResourceNotFoundException {
+public class ScheduleNotFoundException extends ResourceNotFoundException {
+
+    ScheduleNotFoundException() {
```
Comment (xpinjection) [inline scheduleservice/src/main/java/com/kyriba/school/scheduleservice/service/ScheduleNotFoundException.java]: When resource not found there is always search context and it is very helpful when stored in exception message.

File: scheduleservice/src/main/resources/application.properties [hunk 9]
```diff
@@ -0,0 +1,11 @@
+spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.PostgreSQLDialect
+
+spring.jpa.database=POSTGRESQL
+spring.datasource.platform=postgres
+spring.datasource.url=jdbc:postgresql://localhost:5432/schedule
+spring.datasource.username=test
```
Comment (xpinjection) [inline scheduleservice/src/main/resources/application.properties]: Not a good idea to store credentials in default configuration properties.

File: scheduleservice/src/main/resources/application.properties [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+spring.datasource.password=test
+spring.jpa.show-sql=true
```
Comment (xpinjection) [inline scheduleservice/src/main/resources/application.properties]: This option is for dev mode, not default.

File: scheduleservice/src/main/resources/application.properties [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+spring.jpa.generate-ddl=true
+spring.jpa.hibernate.ddl-auto=create-drop
```
Comment (xpinjection) [inline scheduleservice/src/main/resources/application.properties]: Your existing schema will be dropped.

File: scheduleservice/src/test/java/com/kyriba/school/scheduleservice/ScheduleRestControllerIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+	@Autowired
+	private MockMvc mockMvc;
+
+	@MockBean
+	private ScheduleService scheduleService;
+
+	@Test
+	public void get_schedule_by_year_grade_and_letter() throws Exception {
+		int expectedYear = 2019;
```
Comment (xpinjection) [inline scheduleservice/src/test/java/com/kyriba/school/scheduleservice/ScheduleRestControllerIntegrationTest.java]: You could embed them and test will be even more readable.

File: scheduleservice/src/test/java/com/kyriba/school/scheduleservice/dao/AbsenceRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
+import org.springframework.test.annotation.Commit;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+@DBRider
+@ActiveProfiles("test")
+@Commit
```
Comment (xpinjection) [inline scheduleservice/src/test/java/com/kyriba/school/scheduleservice/dao/AbsenceRepositoryTest.java]: If you added flush, then commit is not needed.

File: scheduleservice/src/test/resources/application.properties [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+
+logging.level.org.springframework.web=DEBUG
+
+spring.datasource.driver-class-name=org.testcontainers.jdbc.ContainerDatabaseDriver
+spring.datasource.url=jdbc:tc:postgresql:11-alpine:///schedule
+
+spring.test.database.replace=none
+spring.datasource.username=test
+spring.datasource.password=test
+#spring.jpa.show-sql=true
```
Comment (xpinjection) [inline scheduleservice/src/test/resources/application.properties]: It is usually enabled in tests.

File: scheduleservice/Dockerfile (hunk 1 above)
Comment (vzhuleho) [reply scheduleservice/Dockerfile]: Actually the Dockerfile itself is being placed to the build directory so as scheduleservice.jar. So build directory is used as a source.

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java (hunk 4 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java]: Done

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java (hunk 3 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java]: Done

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java (hunk 2 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/ExternalDataProviderConfiguration.java]: Done

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java (hunk 5 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java]: Done

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java (hunk 6 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/LocalDataProviderConfiguration.java]: Done

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/SchoolClassDataProvider.java (hunk 7 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/dataprovider/SchoolClassDataProvider.java]: Not sure but let it be :) Marked implementation classes as repository

File: scheduleservice/src/main/java/com/kyriba/school/scheduleservice/service/ScheduleNotFoundException.java (hunk 8 above)
Comment (vzhuleho) [reply scheduleservice/src/main/java/com/kyriba/school/scheduleservice/service/ScheduleNotFoundException.java]: Done

File: scheduleservice/src/main/resources/application.properties (hunk 9 above)
Comment (vzhuleho) [reply scheduleservice/src/main/resources/application.properties]: Done

File: scheduleservice/src/main/resources/application.properties (hunk 10 above)
Comment (vzhuleho) [reply scheduleservice/src/main/resources/application.properties]: Done

File: scheduleservice/src/main/resources/application.properties (hunk 11 above)
Comment (vzhuleho) [reply scheduleservice/src/main/resources/application.properties]: Done

File: scheduleservice/src/test/java/com/kyriba/school/scheduleservice/ScheduleRestControllerIntegrationTest.java (hunk 12 above)
Comment (vzhuleho) [reply scheduleservice/src/test/java/com/kyriba/school/scheduleservice/ScheduleRestControllerIntegrationTest.java]: Done

File: scheduleservice/src/test/java/com/kyriba/school/scheduleservice/dao/AbsenceRepositoryTest.java (hunk 13 above)
Comment (vzhuleho) [reply scheduleservice/src/test/java/com/kyriba/school/scheduleservice/dao/AbsenceRepositoryTest.java]: Done

File: scheduleservice/src/test/resources/application.properties (hunk 14 above)
Comment (vzhuleho) [reply scheduleservice/src/test/resources/application.properties]: Done

**Pull request merged**